File: extensions/apexbounty-toolkit/src/PrivilegeMatrixTab.py
# -*- coding: utf-8 -*-
"""
PrivilegeMatrixTab.py - Tab 4: Dynamic Privilege Matrix
Automated BOLA/IDOR detection via background session replays across 4 configured roles.
"""

import threading
from java.awt import BorderLayout, FlowLayout, GridLayout, Color, Component, Dimension
from javax.swing import (
    JPanel, JButton, JLabel, JCheckBox, JTextArea, JScrollPane, JTable, JSplitPane,
    JSeparator, SwingUtilities, BorderFactory, JTabbedPane
)
from javax.swing.table import DefaultTableModel, DefaultTableCellRenderer

from ApexToolkitLogic import PrivilegeMatrixEngine

class StatusColorCellRenderer(DefaultTableCellRenderer):
    """
    Renders role status columns with color codes:
    Green for 200 OK, Red for 401/403, Yellow for Redirects, Orange for 500s.
    """
    def getTableCellRendererComponent(self, table, value, isSelected, hasFocus, row, column):
        cell = DefaultTableCellRenderer.getTableCellRendererComponent(self, table, value, isSelected, hasFocus, row, column)

        # Apply custom coloring to role columns (Admin, User A, User B, Unauth)
        val_str = str(value) if value is not None else ""
        if not isSelected and column >= 2: # Role status columns
            if "200" in val_str or "201" in val_str or "204" in val_str:
                cell.setBackground(Color(200, 245, 200)) # Soft Green
                cell.setForeground(Color.BLACK)
            elif "401" in val_str or "403" in val_str:
                cell.setBackground(Color(255, 210, 210)) # Soft Red
                cell.setForeground(Color.BLACK)
            elif "301" in val_str or "302" in val_str:
                cell.setBackground(Color(255, 255, 200)) # Soft Yellow
                cell.setForeground(Color.BLACK)
            elif "500" in val_str or "502" in val_str:
                cell.setBackground(Color(255, 220, 180)) # Soft Orange
                cell.setForeground(Color.BLACK)
            else:
                cell.setBackground(Color.WHITE)
                cell.setForeground(Color.BLACK)

        return cell


class PrivilegeMatrixTab(object):
    def __init__(self, callbacks, helpers):
        self.callbacks = callbacks
        self.helpers = helpers
        self.is_enabled = False

        self.matrix_records = [] # stores dicts with request/response info for each role

        self._init_ui()

    def _init_ui(self):
        self.panel = JPanel(BorderLayout())

        # Top Control Bar & Configuration Panel
        top_container = JPanel(BorderLayout())

        control_panel = JPanel(FlowLayout(FlowLayout.LEFT, 10, 5))
        lbl_title = JLabel("Dynamic Privilege Matrix (BOLA / IDOR)")
        font = lbl_title.getFont()
        if font:
            lbl_title.setFont(font.deriveFont(font.getStyle() | 1, 14.0))

        self.chk_enable = JCheckBox("Enable Background Matrix Replay", False, actionPerformed=self._on_toggle_enabled)
        self.btn_clear = JButton("Clear Matrix Log", actionPerformed=self._on_clear_log)

        control_panel.add(lbl_title)
        control_panel.add(JSeparator(1))
        control_panel.add(self.chk_enable)
        control_panel.add(self.btn_clear)

        # Role Header Configuration Grid (4 roles: Admin, User A, User B, Unauth)
        grid_panel = JPanel(GridLayout(1, 4, 5, 5))
        grid_panel.setBorder(BorderFactory.createTitledBorder(" Role Auth Headers Configuration "))

        self.txt_admin = JTextArea("Authorization: Bearer ADMIN_TOKEN_HERE\nCookie: session=admin_sess", 3, 20)
        self.txt_usera = JTextArea("Authorization: Bearer USER_A_TOKEN_HERE\nCookie: session=user_a_sess", 3, 20)
        self.txt_userb = JTextArea("Authorization: Bearer USER_B_TOKEN_HERE\nCookie: session=user_b_sess", 3, 20)
        self.txt_unauth = JTextArea("# Unauth: Strips all auth headers", 3, 20)

        grid_panel.add(self._create_role_box("Role 1: Admin", self.txt_admin))
        grid_panel.add(self._create_role_box("Role 2: User A", self.txt_usera))
        grid_panel.add(self._create_role_box("Role 3: User B", self.txt_userb))
        grid_panel.add(self._create_role_box("Role 4: Unauth", self.txt_unauth))

        top_container.add(control_panel, BorderLayout.NORTH)
        top_container.add(grid_panel, BorderLayout.SOUTH)

        # Center Section: Matrix Results Table + Details Viewers
        self.matrix_table_model = DefaultTableModel(
            ["Method", "Path / Endpoint", "Admin Status", "User A Status", "User B Status", "Unauth Status"], 0
        )
        self.matrix_table = JTable(self.matrix_table_model)
        renderer = StatusColorCellRenderer()
        for c in range(6):
            self.matrix_table.getColumnModel().getColumn(c).setCellRenderer(renderer)

        self.matrix_table.getSelectionModel().addListSelectionListener(self._on_row_selected)
        table_scroll = JScrollPane(self.matrix_table)

        # Bottom Detail Tabbed Pane for role responses
        self.role_viewers = {}
        self.detail_tabbed_pane = JTabbedPane()

        roles = ["Admin", "User A", "User B", "Unauth"]
        for role in roles:
            editor = self.callbacks.createMessageEditor(None, False)
            self.role_viewers[role] = editor
            self.detail_tabbed_pane.addTab(role + " Response", editor.getComponent())

        center_split = JSplitPane(JSplitPane.VERTICAL_SPLIT, table_scroll, self.detail_tabbed_pane)
        center_split.setResizeWeight(0.5)

        self.panel.add(top_container, BorderLayout.NORTH)
        self.panel.add(center_split, BorderLayout.CENTER)

    def _create_role_box(self, title, text_area):
        box = JPanel(BorderLayout())
        box.setBorder(BorderFactory.createTitledBorder(title))
        scroll = JScrollPane(text_area)
        box.add(scroll, BorderLayout.CENTER)
        return box

    def get_component(self):
        return self.panel

    def _on_toggle_enabled(self, event):
        self.is_enabled = self.chk_enable.isSelected()

    def _on_clear_log(self, event):
        self.matrix_table_model.setRowCount(0)
        self.matrix_records = []

    def handle_proxy_request(self, message_info):
        """
        Called when a request passes through Burp Proxy.
        """
        if not self.is_enabled:
            return

        http_service = message_info.getHttpService()
        req_bytes = message_info.getRequest()

        if not http_service or not req_bytes:
            return

        req_info = self.helpers.analyzeRequest(http_service, req_bytes)
        url = req_info.getUrl()
        if not url:
            return

        path = url.getPath()
        # Filter static assets
        if any(path.endswith(ext) for ext in ['.js', '.css', '.png', '.jpg', '.jpeg', '.gif', '.svg', '.ico', '.woff', '.woff2']):
            return

        # Spawn background thread to replay across roles
        t = threading.Thread(
            target=self._replay_matrix_thread,
            args=(http_service, req_bytes, req_info.getMethod(), path)
        )
        t.daemon = True
        t.start()

    def _replay_matrix_thread(self, http_service, req_bytes, method, path):
        raw_req = self.helpers.bytesToString(req_bytes)

        roles_config = [
            ("Admin", self.txt_admin.getText()),
            ("User A", self.txt_usera.getText()),
            ("User B", self.txt_userb.getText()),
            ("Unauth", self.txt_unauth.getText())
        ]

        role_statuses = {}
        role_responses = {}

        for role_name, role_headers in roles_config:
            modified_req_str = PrivilegeMatrixEngine.apply_role_headers(raw_req, role_headers)
            modified_req_bytes = self.helpers.stringToBytes(modified_req_str)

            try:
                resp = self.callbacks.makeHttpRequest(http_service, modified_req_bytes)
                status_str = "Error"
                if resp and resp.getResponse():
                    resp_info = self.helpers.analyzeResponse(resp.getResponse())
                    code = resp_info.getStatusCode()
                    status_str = str(code)
                    role_responses[role_name] = resp.getResponse()
                else:
                    role_responses[role_name] = None
                role_statuses[role_name] = status_str
            except Exception as ex:
                role_statuses[role_name] = "Err"
                role_responses[role_name] = None

        record = {
            'method': method,
            'path': path,
            'statuses': role_statuses,
            'responses': role_responses
        }
        self.matrix_records.append(record)

        def add_row():
            self.matrix_table_model.addRow([
                method,
                path,
                role_statuses.get("Admin", "N/A"),
                role_statuses.get("User A", "N/A"),
                role_statuses.get("User B", "N/A"),
                role_statuses.get("Unauth", "N/A")
            ])

        SwingUtilities.invokeLater(add_row)

    def _on_row_selected(self, event):
        if event.getValueIsAdjusting():
            return
        row = self.matrix_table.getSelectedRow()
        if 0 <= row < len(self.matrix_records):
            rec = self.matrix_records[row]
            responses = rec['responses']
            for role_name, editor in self.role_viewers.items():
                resp_bytes = responses.get(role_name)
                if resp_bytes:
                    editor.setMessage(resp_bytes, False)
                else:
                    editor.setMessage(self.helpers.stringToBytes("No Response Recorded"), False)

File: extensions/apexbounty-toolkit/src/main/java/com/example/apexbounty/PrivilegeMatrixTab.java
package com.example.apexbounty;

import burp.api.montoya.MontoyaApi;
import burp.api.montoya.http.message.HttpRequestResponse;
import burp.api.montoya.http.message.requests.HttpRequest;
import burp.api.montoya.http.message.responses.HttpResponse;
import burp.api.montoya.proxy.http.InterceptedRequest;
import burp.api.montoya.ui.editor.HttpResponseEditor;

import javax.swing.*;
import javax.swing.table.DefaultTableCellRenderer;
import javax.swing.table.DefaultTableModel;
import java.awt.*;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CopyOnWriteArrayList;

public class PrivilegeMatrixTab {

    private final MontoyaApi api;
    private final JPanel panel;
    private final JCheckBox chkEnable;
    private final JCheckBox chkStateChanging;
    private final JButton btnClear;

    private final JTextArea txtAdmin;
    private final JTextArea txtUserA;
    private final JTextArea txtUserB;
    private final JTextArea txtUnauth;

    private final DefaultTableModel matrixTableModel;
    private final JTabbedPane detailTabbedPane;
    private final Map<String, HttpResponseEditor> roleViewers = new HashMap<>();

    private final List<MatrixRecord> matrixRecords = new CopyOnWriteArrayList<>();
    private volatile boolean isEnabled = false;

    private static class MatrixRecord {
        final String method;
        final String path;
        final Map<String, String> statuses;
        final Map<String, HttpResponse> responses;

        MatrixRecord(String method, String path, Map<String, String> statuses, Map<String, HttpResponse> responses) {
            this.method = method;
            this.path = path;
            this.statuses = statuses;
            this.responses = responses;
        }
    }

    private static class StatusColorCellRenderer extends DefaultTableCellRenderer {
        @Override
        public Component getTableCellRendererComponent(JTable table, Object value, boolean isSelected, boolean hasFocus, int row, int column) {
            Component cell = super.getTableCellRendererComponent(table, value, isSelected, hasFocus, row, column);
            String valStr = value != null ? String.valueOf(value) : "";

            if (!isSelected && column >= 2) {
                if (valStr.contains("200") || valStr.contains("201") || valStr.contains("204")) {
                    cell.setBackground(new Color(200, 245, 200)); // Soft Green
                    cell.setForeground(Color.BLACK);
                } else if (valStr.contains("401") || valStr.contains("403")) {
                    cell.setBackground(new Color(255, 210, 210)); // Soft Red
                    cell.setForeground(Color.BLACK);
                } else if (valStr.contains("301") || valStr.contains("302")) {
                    cell.setBackground(new Color(255, 255, 200)); // Soft Yellow
                    cell.setForeground(Color.BLACK);
                } else if (valStr.contains("500") || valStr.contains("502") || valStr.contains("Error")) {
                    cell.setBackground(new Color(255, 220, 180)); // Soft Orange
                    cell.setForeground(Color.BLACK);
                } else {
                    cell.setBackground(Color.WHITE);
                    cell.setForeground(Color.BLACK);
                }
            }
            return cell;
        }
    }

    public PrivilegeMatrixTab(MontoyaApi api) {
        this.api = api;
        this.panel = new JPanel(new BorderLayout());

        // Initialize table model first
        matrixTableModel = new DefaultTableModel(
                new Object[]{"Method", "Path / Endpoint", "Admin Status", "User A Status", "User B Status", "Unauth Status"}, 0
        );

        JPanel topContainer = new JPanel(new BorderLayout());

        JPanel controlPanel = new JPanel(new FlowLayout(FlowLayout.LEFT, 10, 5));
        JLabel lblTitle = new JLabel("Dynamic Privilege Matrix (BOLA / IDOR)");
        Font font = lblTitle.getFont();
        if (font != null) {
            lblTitle.setFont(font.deriveFont(Font.BOLD, 14.0f));
        }

        chkEnable = new JCheckBox("Enable Background Matrix Replay", false);
        chkEnable.addActionListener(e -> isEnabled = chkEnable.isSelected());

        chkStateChanging = new JCheckBox("Replay State-Changing Methods (POST/PUT/DELETE)", false);

        btnClear = new JButton("Clear Matrix Log");
        btnClear.addActionListener(e -> {
            matrixTableModel.setRowCount(0);
            matrixRecords.clear();
        });

        controlPanel.add(lblTitle);
        controlPanel.add(new JSeparator(JSeparator.VERTICAL));
        controlPanel.add(chkEnable);
        controlPanel.add(chkStateChanging);
        controlPanel.add(btnClear);

        // Role Config Grid
        JPanel gridPanel = new JPanel(new GridLayout(1, 4, 5, 5));
        gridPanel.setBorder(BorderFactory.createTitledBorder(" Role Auth Headers Configuration "));

        txtAdmin = new JTextArea("Authorization: Bearer ADMIN_TOKEN_HERE\nCookie: session=admin_sess", 3, 20);
        txtUserA = new JTextArea("Authorization: Bearer USER_A_TOKEN_HERE\nCookie: session=user_a_sess", 3, 20);
        txtUserB = new JTextArea("Authorization: Bearer USER_B_TOKEN_HERE\nCookie: session=user_b_sess", 3, 20);
        txtUnauth = new JTextArea("# Unauth: Strips all auth headers", 3, 20);

        gridPanel.add(createRoleBox("Role 1: Admin", txtAdmin));
        gridPanel.add(createRoleBox("Role 2: User A", txtUserA));
        gridPanel.add(createRoleBox("Role 3: User B", txtUserB));
        gridPanel.add(createRoleBox("Role 4: Unauth", txtUnauth));

        topContainer.add(controlPanel, BorderLayout.NORTH);
        topContainer.add(gridPanel, BorderLayout.SOUTH);

        // Results Table
        JTable matrixTable = new JTable(matrixTableModel);
        StatusColorCellRenderer renderer = new StatusColorCellRenderer();
        for (int c = 0; c < 6; c++) {
            matrixTable.getColumnModel().getColumn(c).setCellRenderer(renderer);
        }

        matrixTable.getSelectionModel().addListSelectionListener(e -> {
            if (e.getValueIsAdjusting()) return;
            int row = matrixTable.getSelectedRow();
            if (row >= 0 && row < matrixRecords.size()) {
                MatrixRecord rec = matrixRecords.get(row);
                for (Map.Entry<String, HttpResponseEditor> entry : roleViewers.entrySet()) {
                    String roleName = entry.getKey();
                    HttpResponseEditor editor = entry.getValue();
                    HttpResponse resp = rec.responses.get(roleName);
                    if (resp != null) {
                        editor.setResponse(resp);
                    }
                }
            }
        });

        JScrollPane tableScroll = new JScrollPane(matrixTable);

        detailTabbedPane = new JTabbedPane();
        String[] roles = new String[]{"Admin", "User A", "User B", "Unauth"};
        for (String role : roles) {
            HttpResponseEditor editor = api.userInterface().createHttpResponseEditor();
            roleViewers.put(role, editor);
            detailTabbedPane.addTab(role + " Response", editor.uiComponent());
        }

        JSplitPane centerSplit = new JSplitPane(JSplitPane.VERTICAL_SPLIT, tableScroll, detailTabbedPane);
        centerSplit.setResizeWeight(0.5);

        panel.add(topContainer, BorderLayout.NORTH);
        panel.add(centerSplit, BorderLayout.CENTER);
    }

    private JPanel createRoleBox(String title, JTextArea textArea) {
        JPanel box = new JPanel(new BorderLayout());
        box.setBorder(BorderFactory.createTitledBorder(title));
        box.add(new JScrollPane(textArea), BorderLayout.CENTER);
        return box;
    }

    public Component getComponent() {
        return panel;
    }

    public void handleProxyRequest(InterceptedRequest interceptedRequest) {
        if (!isEnabled || interceptedRequest == null) return;

        HttpRequest request = interceptedRequest;
        if (request.url() == null) return;

        // Target scope check
        if (!api.scope().isInScope(request.url())) {
            return;
        }

        String method = request.method() != null ? request.method().toUpperCase() : "";
        if (isStateChanging(method) && !chkStateChanging.isSelected()) {
            return;
        }

        String path = request.path() != null ? request.path() : "/";
        if (isStaticAsset(path)) {
            return;
        }

        Thread t = new Thread(() -> replayMatrixThread(request, method, path));
        t.setDaemon(true);
        t.start();
    }

    private boolean isStateChanging(String method) {
        return "POST".equals(method) || "PUT".equals(method) || "DELETE".equals(method) || "PATCH".equals(method);
    }

    private boolean isStaticAsset(String path) {
        String lower = path.toLowerCase();
        return lower.endsWith(".js") || lower.endsWith(".css") || lower.endsWith(".png") ||
                lower.endsWith(".jpg") || lower.endsWith(".jpeg") || lower.endsWith(".gif") ||
                lower.endsWith(".svg") || lower.endsWith(".ico") || lower.endsWith(".woff") || lower.endsWith(".woff2");
    }

    private void replayMatrixThread(HttpRequest baseRequest, String method, String path) {
        String[][] rolesConfig = new String[][]{
                {"Admin", txtAdmin.getText()},
                {"User A", txtUserA.getText()},
                {"User B", txtUserB.getText()},
                {"Unauth", txtUnauth.getText()}
        };

        Map<String, String> roleStatuses = new HashMap<>();
        Map<String, HttpResponse> roleResponses = new HashMap<>();

        for (String[] roleConfig : rolesConfig) {
            String roleName = roleConfig[0];
            String roleHeaders = roleConfig[1];

            HttpRequest modifiedReq = ApexToolkitLogic.PrivilegeMatrixEngine.applyRoleHeaders(baseRequest, roleHeaders);

            try {
                HttpRequestResponse reqResp = api.http().sendRequest(modifiedReq);
                HttpResponse resp = reqResp != null ? reqResp.response() : null;
                if (resp != null) {
                    roleStatuses.put(roleName, String.valueOf(resp.statusCode()));
                    roleResponses.put(roleName, resp);
                } else {
                    roleStatuses.put(roleName, "Error");
                    roleResponses.put(roleName, null);
                }
            } catch (Exception ex) {
                roleStatuses.put(roleName, "Error: " + ex.getMessage());
                roleResponses.put(roleName, null);
            }
        }

        MatrixRecord record = new MatrixRecord(method, path, roleStatuses, roleResponses);
        matrixRecords.add(record);

        SwingUtilities.invokeLater(() -> matrixTableModel.addRow(new Object[]{
                method,
                path,
                roleStatuses.getOrDefault("Admin", "N/A"),
                roleStatuses.getOrDefault("User A", "N/A"),
                roleStatuses.getOrDefault("User B", "N/A"),
                roleStatuses.getOrDefault("Unauth", "N/A")
        }));
    }
}
